README.md
# TetraWeb PHP CI Kit with Docker

[![Build Status](https://travis-ci.org/TetraWeb/docker.svg?branch=master)](https://travis-ci.org/TetraWeb/docker)

This repository contains a set of utilities for running PHP tests via [Gitlab CI](https://about.gitlab.com/gitlab-ci/).

These tools provide:

1. [Gitlab runner](https://gitlab.com/gitlab-org/gitlab-ci-multi-runner) deployment script for Ubuntu on VM or metal server. It is not recommended to install runner right on the production system.

2. Set of Docker images for PHP 5.5 - 7.1 based on official Docker PHP images with additional modules and Node.js

The goal of these tools is to automate as much as possible of routine work related to configuring the runner so you can concentrate on writing tests for your code.
Also these tools are trying to be resources savvy, since in most cases huge in-RAM caches are not needed for just running unit tests with some fixtures. So you can use very small VMs for running tests

## Contents of repository
 - [Gitlab-runner bootstrap script](https://github.com/TetraWeb/docker/tree/master/gitlab-runner-vm) for deploying gitlab-runner
 - [PHP Docker images](https://github.com/TetraWeb/docker/tree/master/php) - based on official Docker images, but include addtional modules and also obsolete versions of PHP 5.3 and PHP 5.4
 - [Example projects](https://github.com/TetraWeb/docker/tree/master/examples) - Example PHP projects

## Quick start

1. Install [Gitlab](https://about.gitlab.com/)
1. Get a server (VM or metal) with minimal Ubuntu-14.04 installed. It will be used for the runner
1. Login as root to a server and execute `curl -S https://raw.githubusercontent.com/TetraWeb/docker/master/gitlab-runner-vm/bootstrap.sh | bash` and answer the questions. This script will install docker, Gitlab runner, and configure runner for using these docker images.

Runner is limited to `tetraweb/php:*` images for main container (where your repository is cloned) and any service images `*/*` (secondary containers spinned for services like mysql, redis, etc)

If you want to use the server for also running other images (`ruby` or whatever), you should add another runners to `/etc/gitlab-runner/config.toml`, and DO NOT overwrite `allowed_images = ["tetraweb/php:*"]` for this runner, since it is a potential security breach.

## Requirements
 - Gitlab v`9.0` and later
 - Gitlab runner v`9.0` and later

## Similar projects
 - https://github.com/bobey/docker-gitlab-ci-runner (For old gitlab-ci-runner, misses PHP 5.3)


CHANGELOG.md
# Changelog TetraWeb docker images

`2017-05-28`
 - Add yarn
 - memchache commented out in 7.0 and 7.1 due to compilation errors
 - XDebug updated to 2.5.4 for 7.0 and 7.1

`2017-04-08`
 - Clean up PHP < 5.5
 - Add igbinary

`2016-12-17`
 - [PHP] Added 7.1 image
 - [PHP] Xdebug updated 2.5.0 for PHP 5.5+

`2016-11-02`
 - [PHP] Update xdebug to 2.4.1 for PHP 5.4+
 - [PHP] Add rsync
 - [PHP] Fix compilation issues

`2016-06-20`
 - [PHP] Update xdebug to 2.4.0 for PHP 5.4+

`2016-06-14`
 - [PHP] Nodejs upgraded to 6.x

`2016-06-12`
 - [PHP] Add amqp extension
 - [PHP] Add memcache and redis extensions for PHP7

`2016-04-26`
 - [PHP] Uncomment enabling xdebug from environment variables for PHP7.0 (thanks to @vikpe)

`2016-03-27`
 - [PHP] Remove PHP 5.2 since Debian 6.0 image is not available on Docker hub anymore

`2016-03-13`
 - [PHP] Update PHP7 xdebug version to stable 2.4.0

`2016-03-04`
 - [VM] Disable updating docker images via cron since gitlab-runner is doing that itself

`2016-01-10`
 - [PHP] Add xdebug for PHP 7
 - [PHP] Disable xdebug by default for PHP 5.2

`2016-01-09`
 - [HHVM] Preliminary hhvm release

`2016-01-06`
- [PHP] *WARNING: Images with `-cli` tags are not supported anymore* to decrease docker hub building overhead. So use `tetraweb/php:5.6` instead of `tetraweb/php:5.6-cli`, they are equal
- [PHP] Add missing libicu library for `intl` extension
- [PHP] Add new `mongodb` to 5.4+ images including 7.0



examples/purephp/_tests/test.sh
#!/bin/bash

set -xe

composer self-update

composer --version
phpunit --version

php -i

php/entrypoint.sh
#!/bin/bash
set -e

# With env variable WITH_XDEBUG=1 xdebug extension will be enabled
[ ! -z "$WITH_XDEBUG" ] && docker-php-ext-enable xdebug

# Provide github token if you are using composer a lot in non-interactive mode
# Otherwise one day it will get stuck with request for authorization
# https://github.com/settings/tokens
if [[ ! -z "$COMPOSER_GITHUB" ]]
then
  composer config --global github-oauth.github.com "$COMPOSER_GITHUB"
fi

#
# If $TIMEZONE variable is passed to the image - it will set system timezone 
# and php.ini date.timezone value as well
# Overwise the default system Etc/UTC timezone will be used
#
# Also you can set the php timezone with direct setting it in php.ini 
# within your .gitlab-ci.yml like
# before_script:
# - echo "America/New_York" > /usr/local/etc/php/conf.d/timezone.ini

if [[ ! -z "$TIMEZONE" ]]
then
  echo "$TIMEZONE" > /etc/timezone
  dpkg-reconfigure -f noninteractive tzdata
fi
echo "date.timezone=`cat /etc/timezone`" > /usr/local/etc/php/conf.d/timezone.ini

exec "$@"


gitlab-runner-vm/bootstrap.sh
#!/bin/bash

###############################################################################
#
# This is a bootstrap script for quickly setting up gitlab-runner
# on virtual (or even metal) machine with Ubuntu-14.04
#
# gitlab-runner is configured below to run the separate 
# docker container (runner) for each version of php
#
# TODO:
# System-related variables:
# USE_SWAP="1G" for creating swap
#
# Runner-related variables:
# CI_URL - URL of Gitlab-CI installation
# CI_TOKEN - Token for adding runner. Can be found ($CI_URL/admin/runners)
# CONCURRENT=1 Number of runners which can be executed simultaneously
# COMPOSER_GITHUB="TOKEN" - Github oauth token passed to each container
# TIMEZONE="America/New_York" for setting timezone
#
# Usage:
# 1. Install fresh ubuntu distribution
# 2.1 Non-interactive mode (doesn't require any input):
#     # export CI_URL=http://git.example.com/ci CI_TOKEN=12345abcdef[ OTHER_VAR=VALUE ...]; curl -SL https://raw.githubusercontent.com/TetraWeb/docker/master/gitlab-runner-vm/bootstrap.sh | bash
# OR
# 2.2 Interactive mode (script will ask for all variables to be typed while run)
#     # curl -SL https://raw.githubusercontent.com/TetraWeb/docker/master/gitlab-runner-vm/bootstrap.sh | bash
#
###############################################################################

set -e

create_swap() {
    dd if=/dev/zero of=/swapfile bs=1024 count=1024k
    mkswap /swapfile
    swapon /swapfile
    echo "/swapfile       none    swap    sw      0       0" >> /etc/fstab
    echo 10 | tee /proc/sys/vm/swappiness
    echo vm.swappiness = 10 | tee -a /etc/sysctl.conf
    chown root:root /swapfile
    chmod 0600 /swapfile
}

do_install() {
    while [ -z "$CI_URL" ]; do
        echo -n "Enter URL of your Gitlab CI installation [http://git.example.com/ci]: "
        read CI_URL < /dev/tty
    done

    while [ -z "$CI_TOKEN" ]; do
        echo -n "Enter API key (find it at $CI_URL/admin/runners): "
        read CI_TOKEN < /dev/tty
    done

    if [ -z "$TIMEZONE" ]; then
        echo -n "Timezone name [America/New_York]: "
        read TIMEZONE < /dev/tty
        if [ -z "$TIMEZONE" ]; then
            TIMEZONE="America/New_York"
        fi
        echo "$TIMEZONE" > /etc/timezone
        dpkg-reconfigure -f noninteractive tzdata
    fi

    if [ -z "$COMPOSER_GITHUB" ]; then
        echo -n "Composer Github token (optional): "
        read COMPOSER_GITHUB < /dev/tty
    fi

    if [ ! -z "$USE_SWAP" ]; then
        create_swap
    fi

    if [ -z "$CONCURRENT" ]; then
        echo -n "Number of concurrent processes ($(nproc)): "
        read CONCURRENT < /dev/tty
        if [ -z "$CONCURRENT" ]; then
            CONCURRENT=$(nproc)
        fi
    fi

    export DEBIAN_FRONTEND=noninteractive
    apt-get update && apt-get -y upgrade
    apt-get -y install mc htop ntpdate git curl wget openssh-server

    # Install docker if not found
    docker >/dev/null 2>&1 || { wget -qO- https://get.docker.com/ | sh; }

    # Install multi-runner
    curl -L https://packages.gitlab.com/install/repositories/runner/gitlab-ci-multi-runner/script.deb.sh | sudo bash
    apt-get install gitlab-ci-multi-runner

    # Declare environment variables array
    declare -a ENVVARS

    ENVVARS+=("MYSQL_ALLOW_EMPTY_PASSWORD=1")

    if [ ! -z "$COMPOSER_GITHUB" ]; then
        ENVVARS+=("COMPOSER_GITHUB=$COMPOSER_GITHUB")
    fi
    if [ ! -z "$TIMEZONE" ]; then
        ENVVARS+=("TIMEZONE=$TIMEZONE")
    fi

    gitlab-ci-multi-runner register -n -r "$CI_TOKEN" -u "$CI_URL" --tag-list 'php,mysql' --executor docker \
        --docker-image "tetraweb/php:latest" --docker-allowed-images "tetraweb/php:*" \
        --docker-allowed-services "*" --docker-allowed-services "*/*" $(printf " --env %s" "${ENVVARS[@]}")

    sed -i -- "s/concurrent = 1/concurrent = $CONCURRENT/g" /etc/gitlab-runner/config.toml

    cronjob="#!/bin/bash\n"
#    Automatic updates of images seems to be working in gitlab-runner
#    for phpver in 5.3 5.4 5.5 5.6 7.0
#    do
#        cronjob+="docker pull tetraweb/php:$phpver\n"
#    done

    # Cleanup orphaned images
    cronjob+="docker rmi \$(docker images | grep none | awk '{print \$3}')\n"
    echo -e "$cronjob" > /etc/cron.weekly/docker-update-images
    chmod 755 /etc/cron.weekly/docker-update-images
}

do_install
